### DISCUSSION TITLE
Remote debugging Keycloak

### DISCUSSION THREAD

**Discussion by martinleim** [dasniko/testcontainers-keycloak#116] [Q&A]
I have recently started to use Keycloak Testcontainer for testing a custom Keycloak extension. Registering the extension and starting Keycloak worked like a charm.

When encountering the first bug in my extension, I searched for a way to **remotely debug** within Keycloak and was surprised that I could not find any mention of it within the documentation or the issues and discussion here. After fiddling around with it for some time, I found a solution that involves extending the source code to add a new option `withDebug` that is working for me.

**I am wondering if there is a simpler solution for that. If there is not, is there any interest in bringing this into the project?**

Please see https://github.com/dasniko/testcontainers-keycloak/compare/main...martinleim:testcontainers-keycloak-debug:main for the change set. It can be used in the following way:

1. Enable debugging (without suspending) on a random port. This is a bit cumbersome since you have to find the mapped port and modify your remote debug settings before each run. On the other hand, this probably won't get in the way on a shared environment like a CI server.
```java
@Container
private static final KeycloakContainer keycloak = new KeycloakContainer()
    // ...
    .withDebug();
```
2. Enable debugging on a fixed port. The second parameter controls if the Java process should suspend until the debugger is connected. This allows you to use a static debug port, but is not suitable to run on a CI server. Also, the suspend option requires you to connect the remote debugger two times (as explained in https://github.com/keycloak/keycloak/discussions/12679#discussioncomment-3307999)
```java
@Container
private static final KeycloakContainer keycloak = new KeycloakContainer()
    // ...
    .withDebug(8787, false);

**Comment by dasniko** [comment]
Hi @martinleim, thanks for coming up with this! 👍 
Do you see any chance to provide an automated test for this? I didn't have an idea, yet, but also haven't thought about it that much.

**Reply by dasniko** [reply]
I've created an issue (#117) from this discussion, as this has good chances to become a new feature.
